### PR TITLE
added test_simple_games

### PR DESCRIPTION
**Pull request by yu239-zz**

### CODE REVIEW

File: parl/layers/layer_wrappers.py [hunk 1]
```diff
@@ -533,3 +534,30 @@ def __call__(self, input):
...
+                                dtype,
+                                name=None,
+                                attr=None,
+                                is_bias=False,
+                                default_initializer=None):
+    """
+    Return a function that creates a parameter which cannot be synchronized like those of layers
+
+    This function can be called in Algorithm, so we don't check the caller and require that
+    the variable can be copied.
```
Comment (yu239-zz) [inline parl/layers/layer_wrappers.py]: 'and' -> 'nor'

File: parl/framework/algorithm.py [hunk 2]
```diff
@@ -176,7 +176,7 @@ def get_behavior_model(self):
         """
         return self.model
 
-    def predict(self, inputs, states):
+    def _predict(self, inputs, states):
```
Comment (emailweixu) [inline parl/framework/algorithm.py]: Why do we need to change predict to _predict?

File: parl/algorithm_zoo/simple_algorithms.py [hunk 3]
```diff
@@ -77,6 +79,8 @@ def __init__(self,
                  hyperparas=dict(lr=1e-4),
                  gpu_id=-1,
                  discount_factor=0.99,
+                 estimated_total_num_batches=0,
```
Comment (emailweixu) [inline parl/algorithm_zoo/simple_algorithms.py]: exploration_end_batches might be more accurate

File: parl/framework/algorithm.py (hunk 2 above)
Comment (yu239-zz) [reply parl/framework/algorithm.py]: Because for some algorithms like SimpleQ, they need to insert some preprocessing (e.g., exploration_rate) before doing the prediction(). So they can override the predict() function but inside that they can still call _predict()

File: parl/algorithm_zoo/simple_algorithms.py [hunk 4]
```diff
@@ -87,12 +91,38 @@ def __init__(self,
...
+        if self.explore:
+            counter = self.exploration_counter()
+            ## first compute the current exploration rate
+            rate = 1 - counter * self.exploration_rate_delta
+
+        updated_inputs = dict(exploration_rate=rate)
+        updated_inputs.update(inputs)
+        return self._predict(updated_inputs, states)
+
     def learn(self, inputs, next_inputs, states, next_states, episode_end,
```
Comment (skylian) [inline parl/algorithm_zoo/simple_algorithms.py]: episode_end -> next_episode_end?

File: parl/algorithm_zoo/simple_algorithms.py [hunk 5]
```diff
@@ -87,12 +98,44 @@ def __init__(self,
         self.total_batches = 0
         ## create a reference model
         self.ref_model = deepcopy(model)
+        ## setup exploration
+        self.explore = (exploration_end_batches > 0)
+        if self.explore:
+            self.exploration_counter = layers.create_persistable_variable(
```
Comment (TomorrowIsAnOtherDay) [inline parl/algorithm_zoo/simple_algorithms.py:101]: I understand  that placing exploration info in fluid is extensible for multi-gpu training or distributed training.
but storing current exploration(e.g 0.82) could be a better way? Some times user may want to set current exploration manually, they don't have to compute how many batches it should be .

File: parl/framework/policy_distribution.py [hunk 6]
```diff
@@ -57,10 +66,21 @@ def loglikelihood(self, action):
...
+        assert len(self.dist.shape) == 2
+        return self.dist.shape[1]
+
+    def add_uniform_exploration(self, rate):
+        if not (isinstance(rate, float) and rate == 0):
+            self.dist = self.dist * (1 - rate) + \
+                   1 / float(self.dim) * rate
+
     def loglikelihood(self, action):
         return 0 - layers.cross_entropy(input=self.dist, label=action)
```
Comment (TomorrowIsAnOtherDay) [inline parl/framework/policy_distribution.py]: when using a float tensor as cross_entorpy's label, attr<soft_label> should be True in layer.cross_entorpy API
https://github.com/PaddlePaddle/Paddle/blob/50104f18c732697a6aef5dc73c3d21fe7e538195/paddle/fluid/operators/cross_entropy_op.cc#L122

Comment (yu239-zz) [reply parl/framework/policy_distribution.py]: In this function, action is expected to be an int tensor, because it is a categorical distribution. I will add a type check for action.

File: parl/algorithm_zoo/simple_algorithms.py (hunk 5 above)
Comment (yu239-zz) [reply parl/algorithm_zoo/simple_algorithms.py:101]: Do you mean they want to specify a fixed exploration rate (a number)? Note that this is only an *example* algorithm. The users are encouraged to implemnet their own algorithms based on these examples. This SimpleQ itself can only handle a limited set of simple cases.

File: parl/algorithm_zoo/simple_algorithms.py (hunk 4 above)
Comment (yu239-zz) [reply parl/algorithm_zoo/simple_algorithms.py]: Changed.

File: parl/algorithm_zoo/simple_algorithms.py (hunk 5 above)
Comment (TomorrowIsAnOtherDay) [reply parl/algorithm_zoo/simple_algorithms.py:101]: Yes, sometimes user may train a model with pruning parameters and a small exploration rate.

File: parl/framework/policy_distribution.py [hunk 7]
```diff
@@ -69,30 +89,28 @@ class Deterministic(PolicyDistribution):
...
-        assert False, "You cannot compute likelihood for a deterministic action!"
-
 
-def q_categorical_distribution(q_value, exploration_rate=0.0):
+def q_categorical_distribution(q_value):
     """
     Generate a PolicyDistribution object given a Q value.
-    We first construct a one-hot distribution according to the Q value,
-    and then add an exploration rate to get a probability.
+    We construct a one-hot distribution according to the Q value.
```
Comment (TomorrowIsAnOtherDay) [inline parl/framework/policy_distribution.py:111]: How do you confirm that q_value is a policy distribution? 
Is it necessary to add a softmax layer here?

Comment (yu239-zz) [reply parl/framework/policy_distribution.py:111]: q_value is not a distribution. It is a vector of Q values. This function tries to convert this vector to a one-hot distribution by taking max().

File: parl/algorithm_zoo/simple_algorithms.py (hunk 5 above)
Comment (yu239-zz) [reply parl/algorithm_zoo/simple_algorithms.py:101]: Yes, the scenario you described is possible. However, the user has the chance to inherit the Algorithm class and implement his own algorithm classes. This SimpleQ is just an example.

File: parl/framework/policy_distribution.py (hunk 7 above)
Comment (TomorrowIsAnOtherDay) [reply parl/framework/policy_distribution.py:111]: https://github.com/PaddlePaddle/Paddle/blob/637827a5bc80d6e0a17466e44b087f91601539cb/python/paddle/fluid/layers/nn.py#L971
if it's a one-hot distribution, how could it be used to compute cross_entropy here ? 
https://github.com/yu239/RL/blob/ccadf0db160aa14018212b4bef0748d81c36093f/parl/framework/policy_distribution.py#L88

Comment (yu239-zz) [reply parl/framework/policy_distribution.py:111]: If you look into the SimpleQ code, you will find that this one-hot distribution is only used for prediction but not for learning (at least in the Q-learning algorithm). The reason why we convert Q values to one-hot distribution is to make the prediction procedure the same for both AC and Q: namely, we derive a distribution from Q values and do sampling. The learning code of SimpleQ still uses Bellman Equation.

**Review by skylian — APPROVED** [review]
